source/extensions/filters/http/oauth2/filter.cc

```diff
@@ -173,19 +179,22 @@ std::string encodeHmac(const std::vector<uint8_t>& secret, absl::string_view dom
   return encodeHmacBase64(secret, domain, expires, token, id_token, refresh_token);
 }
 
-// Generates a nonce based on the current time
-std::string generateFixedLengthNonce(TimeSource& time_source) {
-  constexpr size_t length = 16;
-
-  std::string nonce = fmt::format("{}", time_source.systemTime().time_since_epoch().count());
-
-  if (nonce.length() < length) {
-    nonce.append(length - nonce.length(), '0');
-  } else if (nonce.length() > length) {
-    nonce = nonce.substr(0, length);
-  }
+// Generates a non-guessable nonce that can be used to prevent CSRF attacks.
+// The nonce is a base64 encoded SHA256 HMAC generated from the current timestamp and a secret.
+std::string generateNonce(absl::string_view hmac_secret, TimeSource& time_source) {
+  std::vector<uint8_t> hmac_secret_vec(hmac_secret.begin(), hmac_secret.end());
+  std::string timestamp = fmt::format("{}", time_source.systemTime().time_since_epoch().count());
+  return generateHmacBase64(hmac_secret_vec, timestamp);
+}
 
-  return nonce;
+/**
+ * Encodes the state parameter for the OAuth2 flow.
+ * The state parameter is a base64Url encoded JSON object containing the original request URL and a
+ * nonce for CSRF protection.
+ */
+std::string encodeState(const std::string& original_request_url, const std::string& nonce) {
+  std::string json = fmt::format(R"({{"url":"{}","nonce":"{}"}})", original_request_url, nonce);
+  return Base64Url::encode(json.data(), json.size());
```

Comment:
Will the original_request_url contains special character that may need escape?

Reply:
Yeah, good point. the original_request_url needs to be escaped to avoid special characters.